Enable service definitions to be created as YML

rherwig opened the pull request:
### 1. Why is this change necessary?
This change is a quality of life feature for supporting YML services out of the box.

### 2. What does this change do, exactly?
The change defines a protected constant that maps service file names to their respective file loaders. This array is used to look for the respective file and loads the first matching one.
This does not change the default behaviour, since XML services are loaded first and the iteration is cancelled upon first find.

### 3. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

bcremer commented:
Awesome, thanks @andrewisplinghoff :+1:

Commit:
Enable service definitions to be created as YML

keulinho commented on src/Core/Framework/Bundle.php:
```diff
@@ -112,15 +126,25 @@ private function registerEvents(ContainerBuilder $container): void
 
-        $loader = new XmlFileLoader($container, new FileLocator($this->getPath()));
-        $loader->load($containerFilePath);
+        foreach (self::SUPPORTED_CONTAINER_FILES as $path => $loaderClass) {
```
you could use the DelegatingLoader for that one:
```php
$loaderResolver = new LoaderResolver([
    new XmlFileLoader($container, new FileLocator($this->getPath()),
    new YamlFileLoader($container, new FileLocator($this->getPath()),
]);
$delegatingLoader = new DelegatingLoader($loaderResolver);

$delegatingLoader->load($containerFilePath);
```

rherwig replied:
Could you explain the benefit? I would need to iterate either way, just to target the right filepath that the delegating loader needs to resolve a loader, right?

keulinho replied:
i personally would probably go for something like this:
```php
$loaderResolver = new LoaderResolver([
    new XmlFileLoader($container, new FileLocator($this->getPath()),
    new YamlFileLoader($container, new FileLocator($this->getPath()),
]);
$delegatingLoader = new DelegatingLoader($loaderResolver);

foreach (glob($this->getPath() . '/Resources/config/services.*' as $
    $delegatingLoader->load($containerFilePath);
}
```

would be easily extendable, the one problem i see in your implementation is the 
```
$loader = new $loaderClass($container, new FileLocator($this->getPath()));
```
line, because the constructor is in no interface or something, IMHO we should not depend on that these constructor arguments won't change.

The main problem is that the dynamic constructor call is not statically analyzable and fixable.

rherwig replied:
I get your point and it sounds reasonable. I am only a bit concerned, if this is worth the extra allocations. But then again, the services are only loaded once.

I'm gonna rework this and change the PR.

keulinho reviewed (commented):
Thanks @rherwig for your contribution.

I only have a little remark

Commit:
Refactor service loading to use loader resolver.

This change was proposed during the PR discussion and avoids dynamic construction of objects to make the implementation statically analyzable.

mitelg commented:
thanks for your contribution @rherwig :+1: :tada: :blue_heart:
